llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -5325,17 +5321,36 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
     ICmpInst::Predicate Pred2;
     if (match(Op0, m_CombineAnd(m_Instruction(ShiftI),
                                 m_Shr(m_Value(X),
-                                      m_SpecificIntAllowUndef(OpWidth - 1)))) &&
-        match(A, m_ICmp(Pred2, m_Value(Y), m_AllOnes())) &&
-        Pred2 == ICmpInst::ICMP_SGT && X->getType() == Y->getType()) {
+                                      m_SpecificIntAllowUndef(OpWidth - 1))))) {
+      // Test if 2 values have different or same signbits:
+      // (X u>> BitWidth - 1) == zext (Y s> -1) --> (X ^ Y) < 0
+      // (X u>> BitWidth - 1) != zext (Y s> -1) --> (X ^ Y) > -1
+      // (X s>> BitWidth - 1) == sext (Y s> -1) --> (X ^ Y) < 0
+      // (X s>> BitWidth - 1) != sext (Y s> -1) --> (X ^ Y) > -1
       unsigned ExtOpc = ExtI->getOpcode();
       unsigned ShiftOpc = ShiftI->getOpcode();
-      if ((ExtOpc == Instruction::ZExt && ShiftOpc == Instruction::LShr) ||
-          (ExtOpc == Instruction::SExt && ShiftOpc == Instruction::AShr)) {
-        Value *Xor = Builder.CreateXor(X, Y, "xor.signbits");
-        Value *R = (Pred == ICmpInst::ICMP_EQ) ? Builder.CreateIsNeg(Xor)
-                                               : Builder.CreateIsNotNeg(Xor);
-        return replaceInstUsesWith(I, R);
+
+      if (match(A, m_ICmp(Pred2, m_Value(Y), m_AllOnes())) &&
+          Pred2 == ICmpInst::ICMP_SGT && X->getType() == Y->getType()) {
+        if ((ExtOpc == Instruction::ZExt && ShiftOpc == Instruction::LShr) ||
+            (ExtOpc == Instruction::SExt && ShiftOpc == Instruction::AShr)) {
+          Value *Xor = Builder.CreateXor(X, Y, "xor.signbits");
+          Value *R = (Pred == ICmpInst::ICMP_EQ) ? Builder.CreateIsNeg(Xor)
+                                                 : Builder.CreateIsNotNeg(Xor);
+          return replaceInstUsesWith(I, R);
+        }
+      }
+
+      // Transform (X < 0 ==/!= icmp(X)) into (not) xor(X < 0, icmp(X))
+      if (match(A, m_c_ICmp(Pred2, m_Value(X), m_Value())) &&
+          ExtOpc == Instruction::ZExt && ShiftOpc == Instruction::LShr) {
+
+        Value *Xor = Builder.CreateXor(Op0, Op1, "xor.ne");
```

Comment:
You aren't really emitting `xor(X < 0, icmp(X))` here. You are emitting `xor(X u> BitWidth, zext(icmp(X))`
I'd think something like:
`(icmp eq/ne (icmp slt X, 0), icmp(X))` would be simpler tbh.

Reply:
InstCombineAndOrXor would fold `bitwise(X u> BitWidth, zext(icmp(X))` into `(bitwise (icmp slt X, 0), icmp(X))`.
Here I think we emit directly `xor` would be more general (?) for other rhs besides `icmp`.

Reply:
I'd skip the middle-man given that its simpler to emit/think about.